### DISCUSSION TITLE
M/N track logic?

### DISCUSSION THREAD

**Discussion by Huang-Chuan** [dstl/Stone-Soup#432] [Q&A]
Is M/N track logic already available in StoneSoup or it leaves user to implement their own version?

**Comment by sdhiscocks** [comment]
The `MultiMeasurementInitiator` should be what your after, and can also have different components from main tracker.

**Reply by Huang-Chuan** [reply]
I think MultiMeasurementInitiator implements 2/2 which is only part of M/N track logic. In a complete track logic, we have "tentative" and "confirmed" state for potential targets. A potiential target becomes tentative after being detected in consecutive 2 scans and becomes  "confirmed" if detected N times in M following scans.

**Reply by sdhiscocks** [reply]
Thanks for clarifying @Huang-Chuan.

So the logic of the multi-measurement initiator by default is to create a "tentative" track (I think `holding_tracks` is used in the code), after a single measurement (using a [`SimpleMeasurementInitiator`](https://stonesoup.readthedocs.io/en/v0.1b5/stonesoup.initiator.html#stonesoup.initiator.simple.SimpleMeasurementInitiator), and then return the track after two updates (based on [`min_points`](https://stonesoup.readthedocs.io/en/v0.1b5/stonesoup.initiator.html#stonesoup.initiator.simple.MultiMeasurementInitiator.min_points) and [`updates_only`](https://stonesoup.readthedocs.io/en/v0.1b5/stonesoup.initiator.html#stonesoup.initiator.simple.MultiMeasurementInitiator.updates_only)) if it hasn't been deleted.

I think your right that additional components would be required to get _M/N_ logic right like you describe, and there is probably different ways you could go about it.

On requiring two consecutive scans, this could be done by creating a new initiator that just works on finding two consecutive measurements before creating a track...or you could even use another Multi Measurement Initiator (like below), but that may not quite have the logic your after TBH:
```python
initiator = MultiMeasurementInitiator(
    deleter=...,
    data_associator=data_associator,
    updater=updater,
    min_points=...,
    initiator=MultiMeasurementInitiator(
        prior_state=prior,
        measurement_model=None,
        deleter=UpdateTimeStepsDeleter(1),  # Delete if last state not update...
        data_associator=init_data_associator,
        updater=updater,
        min_points=2,  # but yield as soon as two updates.
    )
)
```

But for the _M/N_ aspect, a new deleter is probably required. Something like:
```python
class UpdateCountTimeStepsDeleter(Deleter):
    """Update Count Time Steps based deleter

    Identify tracks for deletion which does have a minimum number of :class:`~.Update`
    states in the last :attr:`time_steps_since_update`.
    """
    min_number_of_updates: int = Property(doc="Minimum number of updates required")
    time_steps: int = Property(doc="Number of time steps to count updates")

    def check_for_deletion(self, track, **kwargs):
        if len(track) < time_steps:  # Track must be long enough
            return False
        return sum(1 for state in track[-self.time_steps:] if isinstance(state, Update)) < min_number_of_updates
```

Bringing that together:

```python
initiator = MultiMeasurementInitiator(
    deleter=UpdateCountTimeStepsDeleter(M, N),
    data_associator=data_associator,
    updater=updater,
    min_points=N,
    update_only=False,
    initiator=MultiMeasurementInitiator(
        prior_state=prior,
        measurement_model=None,
        deleter=UpdateTimeStepsDeleter(1),  # Delete if last state not update...
        data_associator=init_data_associator,
        updater=updater,
        min_points=2,  # but yield as soon as two updates.
    )
)
```

Would be useful to know if this works, and we could wrap it up in a inbuilt component.

**Reply by Huang-Chuan** [reply]
Cool! It never occured to me that initiator could be used in such way(recursive) and I think it works. Thanks for your code. I love the flexibility of StoneSoup framework.
